ProcessLib/GroundwaterFlowProcess.h

```diff
@@ -313,52 +312,51 @@ class GroundwaterFlowProcess : public Process
     {
         DBUG("Postprocessing GroundwaterFlowProcess.");
 
-#ifdef USE_PETSC // Note: this is only a test
-
-        std::vector<PetscScalar>  u(_x->size());
-        _x->getGlobalVector(&u[0]);
-
-        int rank;
-        MPI_Comm_rank(PETSC_COMM_WORLD, &rank);
-        if(rank == 0)
-        {
-            std::string output_file_name = file_name;
-            boost::erase_last(output_file_name, ".vtu");
-            output_file_name.append(".dat");
-            std::ofstream os(output_file_name);
-            os << "SCALARS HEAD double 1"<<endl;
-            os << "LOOKUP_TABLE default"<< endl;
-            for (std::size_t i = 0; i < u.size(); ++i)
-                 os << u[i] << "\n";
-            os.close();
-        }
-#else
         std::string const property_name = "Result";
 
         // Get or create a property vector for results.
-        boost::optional<MeshLib::PropertyVector<double>&> result;
+        boost::optional<MeshLib::PropertyVector<ScalarType>&> result;
         if (_mesh.getProperties().hasPropertyVector(property_name))
         {
             result = _mesh.getProperties().template
-                getPropertyVector<double>(property_name);
+                getPropertyVector<ScalarType>(property_name);
         }
         else
         {
             result = _mesh.getProperties().template
-                createNewPropertyVector<double>(property_name,
+                createNewPropertyVector<ScalarType>(property_name,
                     MeshLib::MeshItemType::Node);
             result->resize(_x->size());
         }
         assert(result && result->size() == _x->size());
 
-        // Copy result
-        for (std::size_t i = 0; i < _x->size(); ++i)
-            (*result)[i] = (*_x)[i];
+        // If in the case of DDC under parallel computation
+        if ( (_x->getRangeEnd() - _x->getRangeBegin()) != _x->size())
+        {
+            std::vector<ScalarType>  u(_x->size());
```

Comment:
It is complex to get a global solution array from distributed Petsc vector data meanwhile to avoid allocate memory for the global array. The reason to have a global solution array is that the solution has to be updated onto the ghost nodes as well. The current approach dose not simply copy vector, it concatenates the local arrays to a global array by a very fast gather function.

For the mesh partitioned by nodes, a straightforward approach without global solution array is to collect ghost entry data among neighbor partitions. The drawbacks of this approach are:
1. Additional memory is also needed for  neighboring ghost nodes.
2. More computation load.

Although the current global array approach involves a little bit more memory than the approach addressed above, its computation speedup is much better.

To improve the memory usage of the current approach in the shared memory node, the application of openmp can be investigated. Also, a software package is available specially for the global array under distrusted machine (http://hpc.pnl.gov/globalarrays/documentation.shtml). However it also involves additional computer resources and memory usage. I would like to add these to the future study list. 

So far, I would like to keep the current implementation.

-------------------------------------------
PS: IS routines are only for index mapping.